ouroboros-network/test/Test/Ouroboros/Network/Testnet/Simulation/Node.hs

```diff
@@ -0,0 +1,446 @@
+      addressToRun :: [RelayAccessPoint]
+                   -> Map Domain [IP]
+                   -> (NtNAddr, RelayAccessPoint)
+                   -> Gen SimArgs
+      addressToRun raps dMap (ntnAddr, rap) = do
+        bgaSlotDuration <- fromInteger <$> choose (0, 100)
+        bgaSeed <- mkStdGen <$> arbitrary
+        quota <- chooseInt (0, 100)
+
+        -- These values approximately correspond to false positive
+        -- thresholds for streaks of empty slots with 99% probability,
+        -- 99.9% probability up to 99.999% probability.
+        -- t = T_s [log (1-Y) / log (1-f)]
+        -- Y = [0.99, 0.999...]
+        --
+        -- T_s = slot length of 1s.
+        -- f = 0.05
+        -- The timeout is randomly picked per bearer to avoid all bearers
+        -- going down at the same time in case of a long streak of empty
+        -- slots. TODO: workaround until peer selection governor.
+        -- Taken from ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
+        mustReplyTimeout <- Just <$> oneof (pure <$> [90, 135, 180, 224, 269])
+
+        stdGen <- mkStdGen <$> arbitrary
+
+        lrp <- genLocalRootPeers raps rap
+
+        peerSelectionTargets <- arbitrary
+        dnsTimeout <- arbitrary
+        dnsLookupDelay <- arbitrary
+
+        return
+         $ SimArgs
+            { saSlot                  = bgaSlotDuration
+            , saSeed                  = bgaSeed
+            , saQuota                 = quota
+            , saMbTime                = mustReplyTimeout
+            , saRelays                = raps
+            , saRng                   = stdGen
+            , saDomainMap             = dMap
+            , saAddr                  = ntnAddr
+            , saLocalRootPeers        = lrp
+            , saLocalSelectionTargets = peerSelectionTargets
+            , saDNSTimeoutSctipt      = dnsTimeout
+            , saDNSLookupDelayScript  = dnsLookupDelay
+            }
+
+-- | Run an arbitrary topology
+diffusion_simulation
+  :: forall m. ( MonadAsync m
+              , MonadFork m
+              , MonadST m
+              , MonadEvaluate m
+              , MonadLabelledSTM m
+              , MonadCatch       m
+              , MonadMask        m
+              , MonadTime        m
+              , MonadTimer       m
+              , MonadThrow  (STM m)
+              , Eq (Async m Void)
+              , forall a. Semigroup a => Semigroup (m a)
+              )
+  => BearerInfo
+  -> DiffusionScript
+  -> m Void
+diffusion_simulation
+  defaultBearerInfo
+  (DiffusionScript args) = withAsyncAll (map (uncurry (runCommands Nothing)) args) $ \nodes -> do
+    (_, x) <- waitAny nodes
+    return x
+  where
+    runCommands
+      :: Maybe (Async m Void
+               , StrictTVar m [(Int, Map RelayAccessPoint PeerAdvertise)])
+      -> SimArgs
+      -> [Commands]
+      -> m Void
+    runCommands Nothing simArgs [] = do
+      threadDelay 3600
+      runCommands Nothing simArgs []
```

Comment:
The outbound governor simulations are written in never terminating style.  The drawback of it is that one has to be careful when printing simulation trace.  I don't object, it's just good to remember.